src/x509_str.c

```diff
@@ -1010,22 +1212,56 @@ WOLFSSL_X509_LOOKUP* wolfSSL_X509_STORE_add_lookup(WOLFSSL_X509_STORE* store,
     return &store->lookup;
 }
 
-int wolfSSL_X509_STORE_add_cert(WOLFSSL_X509_STORE* store, WOLFSSL_X509* x509)
+static int wolfSSL_X509_STORE_add_ca(WOLFSSL_X509_STORE* store,
+                                          WOLFSSL_X509* x509, int type)
 {
     int result = WC_NO_ERR_TRACE(WOLFSSL_FATAL_ERROR);
+    DerBuffer* derCert = NULL;
 
-    WOLFSSL_ENTER("wolfSSL_X509_STORE_add_cert");
-    if (store != NULL && store->cm != NULL && x509 != NULL
-                                                && x509->derCert != NULL) {
-        DerBuffer* derCert = NULL;
-
+    WOLFSSL_ENTER("wolfSSL_X509_STORE_add_ca");
+    if (store != NULL && x509 != NULL && x509->derCert != NULL) {
         result = AllocDer(&derCert, x509->derCert->length,
             x509->derCert->type, NULL);
         if (result == 0) {
             /* AddCA() frees the buffer. */
             XMEMCPY(derCert->buffer,
                             x509->derCert->buffer, x509->derCert->length);
-            result = AddCA(store->cm, &derCert, WOLFSSL_USER_CA, VERIFY);
+            result = AddCA(store->cm, &derCert, type, VERIFY);
+        }
+    }
+
+    return result;
+}
+
+
+int wolfSSL_X509_STORE_add_cert(WOLFSSL_X509_STORE* store, WOLFSSL_X509* x509)
+{
+    int result = WC_NO_ERR_TRACE(WOLFSSL_FATAL_ERROR);
+
+    WOLFSSL_ENTER("wolfSSL_X509_STORE_add_cert");
+    if (store != NULL && store->cm != NULL && x509 != NULL
+                                                && x509->derCert != NULL) {
+        /* Mimic the openssl behavior, must be self signed to be considered
+         * trusted, addCA() internals will do additional checks for
+         * CA=TRUE */
+        if (wolfSSL_X509_NAME_cmp(&x509->issuer, &x509->subject) == 0) {
+            result = wolfSSL_X509_STORE_add_ca(store, x509, WOLFSSL_USER_CA);
+    #if !defined(WOLFSSL_SIGNER_DER_CERT)
+            if (result == WOLFSSL_SUCCESS && store->trusted != NULL) {
+                result = wolfSSL_sk_X509_push(store->trusted, x509);
+                result = (result > 0) ? WOLFSSL_SUCCESS : WOLFSSL_FATAL_ERROR;
+            }
+    #endif
+        }
+        else {
+            if (store->certs != NULL) {
+                result = wolfSSL_sk_X509_push(store->certs, x509);
+                result = (result > 0) ? WOLFSSL_SUCCESS : WOLFSSL_FATAL_ERROR;
+            }
+            else {
+                result = wolfSSL_X509_STORE_add_ca(
```

Comment:
We add the certificate to CA if we don't have a store?
Kinda confusing given the comment above.

Reply:
So this is to support the SSL_CTX current behavior. SSL_CTX has a X509_STORE that is not created with new/free, and allocates internal members of X509_STORE  itself. It also shares a single underlying CM and does operations with it directly. store->certs is allocated in `wolfSSL_X509_STORE_new,` so the only way it can be NULL is the SSL_CTX using it. For those cases I opted to leave the behavior as is, because the SSL_CTX will be expecting the underlying CM to be in certain states after using `wolfSSL_X509_STORE_add_cert`.

Reply:
Add a comment

Reply:
Ok, I do something similar in a few other places as well. Alternatively we could also check on `store->isDynamic` instead.